sklearn/feature_selection/_variance_threshold.py

```diff
@@ -47,7 +48,12 @@ class VarianceThreshold(SelectorMixin, BaseEstimator):
     """
 
     def __init__(self, threshold=0.):
-        self.threshold = threshold
+        if threshold < 0.:
+            warnings.warn("Threshold should be non-negative."
+                          "Automatically set as zero.")
+            self.threshold = 0.
+        else:
+            self.threshold = threshold
```

Comment:
scikit-learn prefers to do [input validation](https://scikit-learn.org/stable/developers/develop.html#instantiation) in `fit`.

On the warning, I would prefer raising an error for `self.threshold < 0`. What do you think?

Reply:
Thank you for the feedback!

I moved validation to `fit`.
And about the warning, I don't have any preference so I changed to raise an error.

I also modified the test to raise an error when the threshold is negative.